### PR TITLE
Multi-dimensional flow model implementation based on SAM implementation

### LINKED ISSUE #23121: Incorporate SAM multi-D kernels into Navier-Stokes module
## Reason
We want to try to consolidate NEAMS thermal hydraulics development in common layers as upstream as possible, e.g. for example in the `navier_stokes` module

## Design
Move SAM multi-D kernels (and related objects such as bcs) into `navier_stokes`

## Impact
Create common layers for INL and ANL development. Let's all work in the same sandbox and create a united product as much as possible

### PR DESCRIPTION
**Pull request by lingzou**
This pull request includes multi-dimensional flow model implementation, both Navier-Stokes model and porous-medium model, based on existing SAM implementation

## Reason

The SAM multi-dimensional flow model is a generic CFD-like implementation which should be made available to public.

## Design

Based on existing SAM implementation.

## Impact

Users have more choices when selecting multi-dimensional flow model implementation.

Closes #23121

### CODE REVIEW

**Comment by lingzou** [comment]
@lindsayad There are a couple of minor things to be fixed but the pull request is ready for review.

**Comment by lindsayad** [comment]
@lingzou is it ok for me to push some commits to your branch? I see some style things I'd like to try to change like `map_var_number` to `mapVarNumber` just to be more consistent with naming in the repo

File: modules/navier_stokes/include/bcs/MDFluidEnergyBC.h [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#pragma once
+
+#include "MDFluidIntegratedBCBase.h"
+#include "Function.h"
+
+/* An integral BC for the energy (temperature) equation  */
```
Comment (lindsayad) [inline modules/navier_stokes/include/bcs/MDFluidEnergyBC.h:15]:
```suggestion
/**
  * An integral BC for the energy (temperature) equation 
  */
```

File: modules/navier_stokes/include/bcs/MDFluidEnergyBC.h [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+public:
+  static InputParameters validParams();
+
+  MDFluidEnergyBC(const InputParameters & parameters);
+  virtual ~MDFluidEnergyBC() {}
+
+protected:
+  virtual Real computeQpResidual();
+  virtual Real computeQpJacobian();
+  virtual Real computeQpOffDiagJacobian(unsigned int jvar);
```
Comment (lindsayad) [inline modules/navier_stokes/include/bcs/MDFluidEnergyBC.h:27]:
```suggestion
  virtual Real computeQpResidual() override;
  virtual Real computeQpJacobian() override;
  virtual Real computeQpOffDiagJacobian(unsigned int jvar) override;
```

File: modules/navier_stokes/include/bcs/MDFluidEnergyDirichletBC.h [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#pragma once
+
+#include "NodalBC.h"
+#include "Function.h"
+
+/* A conditional Dirichlet BC for the energy (temperature) equation  */
```
Comment (lindsayad) [inline modules/navier_stokes/include/bcs/MDFluidEnergyDirichletBC.h:15]:
```suggestion
/**
  * A conditional Dirichlet BC for the energy (temperature) equation  
  */
```

File: modules/navier_stokes/include/bcs/MDFluidEnergyDirichletBC.h [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+public:
+  static InputParameters validParams();
+
+  MDFluidEnergyDirichletBC(const InputParameters & parameters);
+  virtual ~MDFluidEnergyDirichletBC() {}
+
+protected:
+  virtual Real computeQpResidual();
+  virtual Real computeQpJacobian();
+  virtual Real computeQpOffDiagJacobian(unsigned int jvar);
```
Comment (lindsayad) [inline modules/navier_stokes/include/bcs/MDFluidEnergyDirichletBC.h:27]:
```suggestion
  virtual Real computeQpResidual() override;
  virtual Real computeQpJacobian() override;
  virtual Real computeQpOffDiagJacobian(unsigned int jvar) override;
```

File: modules/navier_stokes/include/bcs/MDFluidEnergyDirichletBC.h [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+
+  bool shouldApply();
```
Comment (lindsayad) [inline modules/navier_stokes/include/bcs/MDFluidEnergyDirichletBC.h:29]:
```suggestion
  bool shouldApply() override;
```

File: modules/navier_stokes/include/bcs/MDFluidIntegratedBCBase.h [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+  unsigned _w_vel_var_number;
+  unsigned _pressure_var_number;
+  unsigned _temperature_var_number;
+
+  bool _has_porosity;
+  const VariableValue & _porosity;
+
+  // Helper function for mapping Moose variable numberings into
+  // the "canonical" numbering for the porous medium equations.
+  unsigned map_var_number(unsigned var);
```
Comment (lindsayad) [inline modules/navier_stokes/include/bcs/MDFluidIntegratedBCBase.h:56]:
```suggestion
  unsigned mapVarNumber(unsigned var);
```

File: modules/navier_stokes/include/bcs/MDFluidMassBC.h [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#pragma once
+
+#include "MDFluidIntegratedBCBase.h"
+#include "Function.h"
+
+/* A specific BC for the mass (pressure) equation  */
```
Comment (lindsayad) [inline modules/navier_stokes/include/bcs/MDFluidMassBC.h:15]:
```suggestion
/**
  * A specific BC for the mass (pressure) equation 
  */
```

File: modules/navier_stokes/include/bcs/MDFluidMassBC.h [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+public:
+  static InputParameters validParams();
+
+  MDFluidMassBC(const InputParameters & parameters);
+  virtual ~MDFluidMassBC() {}
+
+protected:
+  virtual Real computeQpResidual();
+  virtual Real computeQpJacobian();
+  virtual Real computeQpOffDiagJacobian(unsigned int jvar);
```
Comment (lindsayad) [inline modules/navier_stokes/include/bcs/MDFluidMassBC.h:27]:
```suggestion
  virtual Real computeQpResidual() override;
  virtual Real computeQpJacobian() override;
  virtual Real computeQpOffDiagJacobian(unsigned int jvar) override;
```

File: modules/navier_stokes/include/bcs/MDFluidMomentumBC.h [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+{
+public:
+  static InputParameters validParams();
+
+  MDFluidMomentumBC(const InputParameters & parameters);
+
+protected:
+  virtual Real computeQpResidual();
+  virtual Real computeQpJacobian();
+  virtual Real computeQpOffDiagJacobian(unsigned int jvar);
```
Comment (lindsayad) [inline modules/navier_stokes/include/bcs/MDFluidMomentumBC.h:25]:
```suggestion
  virtual Real computeQpResidual() override;
  virtual Real computeQpJacobian() override;
  virtual Real computeQpOffDiagJacobian(unsigned int jvar) override;
```

File: modules/navier_stokes/include/kernels/MDFluidEnergyKernel.h [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+
+#pragma once
+
+#include "MDFluidKernelStabilization.h"
+
+#include "Function.h"
+
+// Forward Declarations
+class MDFluidEnergyKernel;
+
```
Comment (lindsayad) [inline modules/navier_stokes/include/kernels/MDFluidEnergyKernel.h:18]:
```suggestion
```

File: modules/navier_stokes/include/kernels/MDFluidEnergyKernel.h [hunk 11]
```diff
@@ -0,0 +1,43 @@
...
+// The spatial part of the 3D energy conservation for fluid flow
```
Comment (lindsayad) [inline modules/navier_stokes/include/kernels/MDFluidEnergyKernel.h:19]:
```suggestion
/**
  * The spatial part of the 3D energy conservation for fluid flow
  */
```

File: modules/navier_stokes/include/kernels/MDFluidKernelBase.h [hunk 12]
```diff
@@ -0,0 +1,85 @@
...
+  MDFluidKernelBase(const InputParameters & parameters);
+  virtual ~MDFluidKernelBase() {}
+
+protected:
+  Real velocityDiv() { return _grad_u_vel[_qp](0) + _grad_v_vel[_qp](1) + _grad_w_vel[_qp](2); }
+  RealVectorValue velocityDot()
+  {
+    RealVectorValue vec_vel_dot(_u_vel_dot[_qp], _v_vel_dot[_qp], _w_vel_dot[_qp]);
+    return vec_vel_dot;
+  }
```
Comment (lindsayad) [inline modules/navier_stokes/include/kernels/MDFluidKernelBase.h:34]: In general I try to encourage people (including myself who's often a guilty party) to move multi-line definitions out of the class definition, e.g. we could make this inline in the header outside of the class definition

File: modules/navier_stokes/include/kernels/MDFluidKernelBase.h [hunk 13]
```diff
@@ -0,0 +1,85 @@
...
+  // Material properties
+  const MaterialProperty<RealTensorValue> & _viscous_stress_tensor;
+  const MaterialProperty<Real> & _dynamic_viscosity;
+  const MaterialProperty<Real> & _turbulence_viscosity;
+  const MaterialProperty<RealTensorValue> & _inertia_resistance_coeff;
+  const MaterialProperty<RealTensorValue> & _viscous_resistance_coeff;
+
+  // Helper function for mapping Moose variable numberings into
+  // the "canonical" numbering for the porous medium equations.
+  unsigned map_var_number(unsigned var);
```
Comment (lindsayad) [inline modules/navier_stokes/include/kernels/MDFluidKernelBase.h:80]:
```suggestion
  /**
    * Helper function for mapping Moose variable numberings into
    * the "canonical" numbering for the porous medium equations.
    */
  unsigned int mapVarNumber(unsigned int var) const;
```

File: modules/navier_stokes/include/kernels/MDFluidKernelStabilization.h [hunk 14]
```diff
@@ -0,0 +1,33 @@
...
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#pragma once
+
+#include "MDFluidKernelBase.h"
+
+/* Base class for stabilization kernels. */
```
Comment (lindsayad) [inline modules/navier_stokes/include/kernels/MDFluidKernelStabilization.h:14]:
```suggestion
/**
  * Base class for stabilization kernels.
  */
```

File: modules/navier_stokes/include/kernels/MDFluidMassKernel.h [hunk 15]
```diff
@@ -0,0 +1,31 @@
...
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#pragma once
+
+#include "MDFluidKernelStabilization.h"
+
+// The spatial part of the 3D mass conservation for fluid flow
```
Comment (lindsayad) [inline modules/navier_stokes/include/kernels/MDFluidMassKernel.h:14]:
```suggestion
/**
  * The spatial part of the 3D mass conservation for fluid flow
  */
```

File: modules/navier_stokes/include/kernels/MDFluidMomentumKernel.h [hunk 16]
```diff
@@ -0,0 +1,31 @@
...
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#pragma once
+
+#include "MDFluidKernelStabilization.h"
+
+// The spatial part of the 3D momentum conservation for fluid flow
```
Comment (lindsayad) [inline modules/navier_stokes/include/kernels/MDFluidMomentumKernel.h:14]:
```suggestion
/**
  * The spatial part of the 3D momentum conservation for fluid flow
  */
```

File: modules/navier_stokes/include/kernels/PMFluidPressureTimeDerivative.h [hunk 17]
```diff
@@ -0,0 +1,36 @@
...
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#pragma once
+
+#include "TimeKernel.h"
+#include "SinglePhaseFluidProperties.h"
+
+// The transient term of the porous-media mass conservation equation
```
Comment (lindsayad) [inline modules/navier_stokes/include/kernels/PMFluidPressureTimeDerivative.h:15]:
```suggestion
/**
  * The transient term of the porous-media mass conservation equation
  */
```

File: modules/navier_stokes/include/materials/MDFluidMaterial.h [hunk 18]
```diff
@@ -0,0 +1,76 @@
...
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#pragma once
+
+#include "Function.h"
+#include "Material.h"
+#include "SinglePhaseFluidProperties.h"
+
+/* Fluid materials for 3D fluid model */
```
Comment (lindsayad) [inline modules/navier_stokes/include/materials/MDFluidMaterial.h:16]:
```suggestion
/**
  * Fluid materials for 3D fluid model
  */
```

File: modules/navier_stokes/src/auxkernels/NSDensityAux.C [hunk 19]
```diff
@@ -0,0 +1,39 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "NSDensityAux.h"
```
Comment (lindsayad) [inline modules/navier_stokes/src/auxkernels/NSDensityAux.C:10]:
```suggestion
#include "NSDensityAux.h"
#include "NS.h"
```

File: modules/navier_stokes/src/auxkernels/NSDensityAux.C [hunk 20]
```diff
@@ -0,0 +1,39 @@
...
+registerMooseObject("NavierStokesApp", NSDensityAux);
+
+InputParameters
+NSDensityAux::validParams()
+{
+  InputParameters params = AuxKernel::validParams();
+
+  // Coupled variables
+  params.addRequiredCoupledVar("pressure", "Coupled fluid pressure variable");
+  params.addRequiredCoupledVar("temperature", "Coupled fluid temperature variable");
```
Comment (lindsayad) [inline modules/navier_stokes/src/auxkernels/NSDensityAux.C:21]:
```suggestion
  params.addRequiredCoupledVar(NS::pressure, "Coupled fluid pressure variable");
  params.addRequiredCoupledVar(NS::temperature, "Coupled fluid temperature variable");
```

**Review by lindsayad — DISMISSED** [review]
Some preliminary review through `include`. @lingzou I'd be happy to push commits to address these kind of cosmetic changes to help reduce time burden on you. Just let me know

**Comment by lingzou** [comment]
> 

@lindsayad Please go ahead do the necessary changes. Let me know if I have to do anything on my end, e.g., giving you access to the repo, things like that.

**Comment by lindsayad** [comment]
Sounds good!

**Comment by lindsayad** [comment]
@lingzou would it be possible to add a couple more tests? Currently it looks like the following objects aren't used:
- FluidWallMomentumBC
- MDFluidEnergyBC
- MDFluidEnergyDirichletBC
- MDFluidMomentumBC
- MDMomentumFreeSlipBC
I looked at the possibility of porting some of the existing tests in SAM that use these objects but it looks like they use the non-porous time derivative kernels so I would have to port those as well ... and it looks like some of those time derivative kernels use EOS APIs that don't have drop-in replacements with `SinglePhaseFluidProperties`, e.g. `enthalpy_from_T`. We have `h_from_p_T` but I didn't want to go too far down the rabbit hole of porting those kernels and changing to `SinglePhaseFluidProperties` and changing to `p_T` lookups ... as you may very well want to continue to use objects like `PBSodiumEquationOfState`

**Comment by lingzou** [comment]
> @lingzou would it be possible to add a couple more tests? Currently it looks like the following objects aren't used:
> 
> * FluidWallMomentumBC
> * MDFluidEnergyBC
> * MDFluidEnergyDirichletBC
> * MDFluidMomentumBC
> * MDMomentumFreeSlipBC
> 
> I looked at the possibility of porting some of the existing tests in SAM that use these objects but it looks like they use the non-porous time derivative kernels so I would have to port those as well ... and it looks like some of those time derivative kernels use EOS APIs that don't have drop-in replacements with `SinglePhaseFluidProperties`, e.g. `enthalpy_from_T`. We have `h_from_p_T` but I didn't want to go too far down the rabbit hole of porting those kernels and changing to `SinglePhaseFluidProperties` and changing to `p_T` lookups ... as you may very well want to continue to use objects like `PBSodiumEquationOfState`

@lindsayad thanks for keeping working on the PR. It is probably the best for me to add/port additional tests to cover these BCs. It shouldn't take long to do so, and we will let Elia understand that we are still working on it.

**Comment by lindsayad** [comment]
That would be great, thanks!

**Comment by lindsayad** [comment]
Ok @GiudGiud, @grmnptr I have commits on this PR so for SQA I cannot approve/merge it. Can one of you guys review this? Note that we cannot change the functionality of these objects as we need to maintain SAM compatibility and I'm not going to spend time on consolidation of the FE objects before this is merged per Elia's request

**Review by GiudGiud — COMMENTED** [review]
NSDensityAux should be deleted we have FluidDensityAux in the FP module
FluidWallMomentumBC is missing a prefix

Please link issues for:
- missing docs & docstrings
- issues attached to the original code
- missing Jacobian terms for that test that uses the FDP

Can we rename MD and PM? To INSFE and PINSFE most likely?

**Comment by lindsayad** [comment]
> Can we rename MD and PM? To INSFE and PINSFE most likely?

We could potentially do aliases. If we renamed we would break SAM

**Comment by GiudGiud** [comment]
Aliasing seems fine for a first step.

**Comment by lindsayad** [comment]
Ok aliasing doesn't seem like the correct solution. That only makes sense if you want the name in the `Registry` (and the name usable in the input file) to be different than the actual class name. It's not to support multiple allowable names in an input file. So we could deprecate the current names in favor of new names ... then both names would be usable in the input file but this would be denoting that we essentially want our users to adopt a new name eventually. @lingzou is this something that you think is acceptable?

**Comment by lingzou** [comment]
> ... that we essentially want our users to adopt a new name eventually. @lingzou is this something that you think is acceptable?

@lindsayad Yes, that will be fine, as long as we provide a time window for end users to adopt the new syntax/names. [I am at-ing @lingzou-anl so my ANL email receives update].

File: modules/navier_stokes/doc/content/source/materials/MDFluidMaterial.md [hunk 21]
```diff
@@ -0,0 +1,23 @@
+# MDFluidMaterial
```
Comment (GiudGiud) [inline modules/navier_stokes/doc/content/source/materials/MDFluidMaterial.md:1]: rename?

File: modules/navier_stokes/include/bcs/INSFEFluidWallMomentumBC.h [hunk 22]
```diff
@@ -0,0 +1,29 @@
...
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#pragma once
+
+#include "INSFEFluidIntegratedBCBase.h"
+
+class INSFEFluidWallMomentumBC : public INSFEFluidIntegratedBCBase
```
Comment (GiudGiud) [inline modules/navier_stokes/include/bcs/INSFEFluidWallMomentumBC.h:14]: docstring

File: modules/navier_stokes/include/bcs/INSFEFluidMomentumBC.h [hunk 23]
```diff
@@ -0,0 +1,43 @@
...
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#pragma once
+
+#include "INSFEFluidIntegratedBCBase.h"
+#include "Function.h"
+
+class INSFEFluidMomentumBC : public INSFEFluidIntegratedBCBase
```
Comment (GiudGiud) [inline modules/navier_stokes/include/bcs/INSFEFluidMomentumBC.h:15]: same

File: modules/navier_stokes/include/kernels/PINSFEFluidTemperatureTimeDerivative.h [hunk 24]
```diff
@@ -0,0 +1,32 @@
...
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#pragma once
+
+#include "TimeDerivative.h"
+#include "SinglePhaseFluidProperties.h"
+
+class PINSFEFluidTemperatureTimeDerivative : public TimeDerivative
```
Comment (GiudGiud) [inline modules/navier_stokes/include/kernels/PINSFEFluidTemperatureTimeDerivative.h:15]: docstring

File: modules/navier_stokes/include/kernels/PINSFEFluidVelocityTimeDerivative.h [hunk 25]
```diff
@@ -0,0 +1,32 @@
...
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#pragma once
+
+#include "TimeDerivative.h"
+#include "SinglePhaseFluidProperties.h"
+
+class PINSFEFluidVelocityTimeDerivative : public TimeDerivative
```
Comment (GiudGiud) [inline modules/navier_stokes/include/kernels/PINSFEFluidVelocityTimeDerivative.h:15]: same

File: modules/navier_stokes/include/materials/GenericPorousMediumMaterial.h [hunk 26]
```diff
@@ -0,0 +1,31 @@
...
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#pragma once
+
+#include "MDFluidMaterial.h"
+
+class GenericPorousMediumMaterial : public MDFluidMaterial
```
Comment (GiudGiud) [inline modules/navier_stokes/include/materials/GenericPorousMediumMaterial.h:14]: GenericPINSFEMaterial ?

File: modules/navier_stokes/doc/content/modules/navier_stokes/index.md [hunk 27]
```diff
@@ -27,6 +27,8 @@ Here we give a brief tabular summary of the Navier-Stokes implementations:
 | ------     | --------   | -------------------- | ------------------ | ----------------  | ------ | ------------------                |
 | INS        | Hand-coded | incompressible       | None               | Not porous        | CGFE   | SUPG                              |
 | INSAD      | AD         | incompressible       | Smagorinsky        | Not porous        | CGFE   | SUPG                              |
+| INSFE      | Hand-coded | incompressible       | None               | Not porous        | CGFE   | SUPG                              |
+| PINSFE     | Hand-coded | incompressible       | None               | porous            | CGFE   | SUPG                              |
```
Comment (GiudGiud) [inline modules/navier_stokes/doc/content/modules/navier_stokes/index.md:31]: both have mixing length & seemingly others, see the MDFluidMaterial

File: modules/navier_stokes/test/tests/finite_element/pins/expansion-channel/tests [hunk 28]
```diff
@@ -0,0 +1,12 @@
+[Tests]
+  design = 'modules/navier_stokes/index.md'
+  [slip_wall]
+    type = 'Exodiff'
+    input = 'expansion-channel-slip-wall.i'
+    exodiff = 'expansion-channel-slip-wall_out.e'
+    abs_zero = 1e-6
+    issues = '#23121'
+    requirement = 'We shall be able to solve a porous flow using slip-wall boundary condition.'
```
Comment (GiudGiud) [inline modules/navier_stokes/test/tests/finite_element/pins/expansion-channel/tests:9]:
```suggestion
    requirement = 'The system shall be able to solve a porous flow using slip-wall boundary condition.'
```

File: modules/navier_stokes/test/tests/finite_element/pins/channel-flow/tests [hunk 29]
```diff
@@ -0,0 +1,27 @@
+[Tests]
+  design = 'modules/navier_stokes/index.md'
+  [./pm_friction]
+    type = 'Exodiff'
+    input = 'pm_friction.i'
+    exodiff = 'pm_friction_out.e'
+    abs_zero = 1e-6
+    issues = '#23121'
+    requirement = 'We shall be able to solve a porous flow with pressure drop due to viscous and inertia frictions.'
```
Comment (GiudGiud) [inline modules/navier_stokes/test/tests/finite_element/pins/channel-flow/tests:9]:
```suggestion
    requirement = 'The system shall be able to solve a porous flow with pressure drop due to viscous and inertia frictions.'
```

File: modules/navier_stokes/test/tests/finite_element/pins/channel-flow/tests [hunk 30]
```diff
@@ -0,0 +1,27 @@
...
+  [../]
+  [./pm_heat_source]
+    type = 'Exodiff'
+    input = 'pm_heat_source.i'
+    exodiff = 'pm_heat_source_out.e'
+    abs_zero = 1e-6
+    issues = '#23121'
+    requirement = 'We shall be able to solve a porous flow with internal heating source.'
```
Comment (GiudGiud) [inline modules/navier_stokes/test/tests/finite_element/pins/channel-flow/tests:17]:
```suggestion
    requirement = 'The system shall be able to solve a porous flow with internal heating source.'
```

File: modules/navier_stokes/test/tests/finite_element/pins/channel-flow/tests [hunk 31]
```diff
@@ -0,0 +1,27 @@
...
+  [../]
+  [./pm_reverse_flow]
+    type = 'Exodiff'
+    input = 'pm_reverse_flow.i'
+    exodiff = 'pm_reverse_flow_out.e'
+    abs_zero = 1e-6
+    issues = '#23121'
+    requirement = 'We shall be able to solve a porous flow when reverse flow happens.'
```
Comment (GiudGiud) [inline modules/navier_stokes/test/tests/finite_element/pins/channel-flow/tests:25]:
```suggestion
    requirement = 'The system shall be able to solve a porous flow when reverse flow happens.'
```

File: modules/navier_stokes/test/tests/finite_element/ins/lid_driven/tests [hunk 32]
```diff
@@ -8,6 +8,14 @@
     issues = '#000'
     requirement = 'We shall be able to solve a canonical lid-driven problem without stabilization, using mixed order finite elements for velocity and pressure.'
   [../]
+  [./lid_driven_md]
+    type = 'Exodiff'
+    input = 'lid_driven_md.i'
+    exodiff = 'lid_driven_md_out.e'
+    custom_cmp = 'lid_driven.cmp'
+    issues = '#23121'
+    requirement = 'We shall be able to solve a canonical lid-driven problem using same order variables and the PSPG/SUPG stabilization scheme.'
```
Comment (GiudGiud) [inline modules/navier_stokes/test/tests/finite_element/ins/lid_driven/tests:17]:
```suggestion
    requirement = 'The system shall be able to solve a canonical lid-driven problem using same order variables and the PSPG/SUPG stabilization scheme.'
```

**Review by GiudGiud — COMMENTED** [review]
ok let s get this in after these. 

In terms of coverage we could add two turbulence tests. 
I m ok with adding them later
Same for all object docs

**Review by GiudGiud — DISMISSED** [review]
5 last comments can be done in a follow up they are minor

**Comment by lindsayad** [comment]
Sorry @GiudGiud I pushed another commit with my airplane work, so will need another review

**Comment by lindsayad** [comment]
I tried to address the other 5 conversations

**Review by GiudGiud — APPROVED** [review]
